src/Particle/ParticleIO/XMLParticleIO.cpp

```diff
@@ -244,6 +245,18 @@ bool XMLParticleParser::readXML(xmlNodePtr cur)
             {
               count_group_size++;
               map_storage_to_input[storage_index++] = iat;
+              if (iat != storage_index - 1 && outputManager.isHighActive())
+              {
+                if (!ungrouped_message_printed)
+                {
+                  app_log() << "  Input particle set is not grouped by species.  Remapping particle position indices "
+                               "internally."
+                            << std::endl;
+                  app_log() << "    Species : input particle index -> internal particle index" << std::endl;
+                  ungrouped_message_printed = true;
+                }
+                app_log() << "    " << group_species_name << " : " << iat << " -> " << storage_index - 1 << std::endl;
+              }
```

Comment:
This is printout is sensitive to the ordering of element groups.
It is better to only set `input_ungrouped` to true here and then after the ig loop, handling the printout based on map_storage_to_input.

PS:
1) it is better to consume storage_index before storage_index++ to avoid adding confusing -1.
2) is isHighActive() necessary, should app_log() have been set accordingly?

Reply:
Moved the output code to come after the loop.

1. I would prefer to put code in order of importance - this code is less important than the code that actually stores the reordering.  But the if-test is now before the increment.
2. The isHighActive is not necessary, it is implicit in app_log.